### PR TITLE
Support parse `DateTime64` and timezone in joda syntax

### PR DESCRIPTION
**Pull request by KevinyhZou**
When use function `parseDateTimeInJodaSyntaxOrNull` to parse datetime, it can not handle the following suitations (contains microsecond or timezone):
`select parseDateTimeInJodaSyntaxOrNull('2024-10-09 10:30:30.123', 'yyyy-MM-dd HH:mm:ss.SSS')`  -- contain milliseond
`select parseDateTimeInJodaSyntaxOrNull('2024-10-09 10:30:30.123-0800', 'yyyy-MM-dd HH:mm:ss.SSSZ') ` 
`select parseDateTimeInJodaSyntaxOrNull('2024-10-09 10:30:30.123Asia/Shanghai', 'yyyy-MM-dd HH:mm:ss.SSSz')`  -- which contains timezone

and we encounter these problems in gluten which based on `parseDateTimeInJodaSyntaxOrNull` function, related to https://github.com/apache/incubator-gluten/issues/7564

So we introduce a new function named `parseDateTime64InJodaSyntaxOrNull` to support parse microsecond and timezone.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support parse `DateTime64` for microseond and timezone in joda syntax

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 6c1a709d304d9a2f3145c70cd233e6b3fcb066ec with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70737/6c1a709d304d9a2f3145c70cd233e6b3fcb066ec/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/7e1f7eca66596b0bcabe117d6699033203e2668c/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Functions/parseDateTime.cpp [hunk 1]
```diff
@@ -2097,12 +2222,18 @@ namespace
...
+        static constexpr auto name = "parseDateTime64InJodaSyntaxOrNull";
+    };
+
     using FunctionParseDateTime = FunctionParseDateTimeImpl<NameParseDateTime, ParseSyntax::MySQL, ErrorHandling::Exception>;
     using FunctionParseDateTimeOrZero = FunctionParseDateTimeImpl<NameParseDateTimeOrZero, ParseSyntax::MySQL, ErrorHandling::Zero>;
     using FunctionParseDateTimeOrNull = FunctionParseDateTimeImpl<NameParseDateTimeOrNull, ParseSyntax::MySQL, ErrorHandling::Null>;
     using FunctionParseDateTimeInJodaSyntax = FunctionParseDateTimeImpl<NameParseDateTimeInJodaSyntax, ParseSyntax::Joda, ErrorHandling::Exception>;
     using FunctionParseDateTimeInJodaSyntaxOrZero = FunctionParseDateTimeImpl<NameParseDateTimeInJodaSyntaxOrZero, ParseSyntax::Joda, ErrorHandling::Zero>;
     using FunctionParseDateTimeInJodaSyntaxOrNull = FunctionParseDateTimeImpl<NameParseDateTimeInJodaSyntaxOrNull, ParseSyntax::Joda, ErrorHandling::Null>;
+    using FunctionParseDateTime64InJodaSyntaxOrNull = FunctionParseDateTimeImpl<NameParseDateTime64InJodaSyntaxOrNull, ParseSyntax::Joda, ErrorHandling::Null, true>;
```
Comment (hanfei1991) [inline src/Functions/parseDateTime.cpp:2236]: Shall we also support `NameParseDateTime64InJodaSyntax` and `NameParseDateTime64InJodaSyntaxOrZero`?

Comment (KevinyhZou) [reply src/Functions/parseDateTime.cpp:2248]: done

File: tests/queries/0_stateless/03252_parse_datetime64_in_joda_syntax.reference [hunk 2]
```diff
@@ -0,0 +1,14 @@
+2024-10-09 10:30:10.123
+2024-10-09 10:30:10.123456
+2024-10-10 02:30:10.123456
+2024-10-09 10:24:27.123456
```
Comment (hanfei1991) [inline tests/queries/0_stateless/03252_parse_datetime64_in_joda_syntax.reference:4]: Have not read the code thoroughly, just a question:
Why `select parseDateTime64InJodaSyntax('2024-10-09 10:30:10.123456Asia/Shanghai', 'yyyy-MM-dd HH:mm:ss.SSSSSSz');` get result `2024-10-09 10:24:27.123456`

Comment (KevinyhZou) [reply tests/queries/0_stateless/03252_parse_datetime64_in_joda_syntax.reference]: The time offset computing between timezones is wrong, and now it is fixed.

**Comment by KevinyhZou** [comment]
cc @hanfei1991

**Review by hanfei1991 — APPROVED** [review]

File: src/Functions/parseDateTime.cpp [hunk 3]
```diff
@@ -598,13 +610,71 @@ namespace
...
+                date_type = std::make_shared<DataTypeDateTime>(time_zone_name);
             if (error_handling == ErrorHandling::Null)
                 return std::make_shared<DataTypeNullable>(date_type);
             return date_type;
         }
 
-        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & /*result_type*/, size_t input_rows_count) const override
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+        {
+            ColumnUInt8::MutablePtr col_null_map;
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:649]: L. 649-641, l. 658-661 and l. 668-671 are common and should be moved into `execute2()`.

File: src/Functions/parseDateTime.cpp [hunk 4]
```diff
@@ -598,13 +610,71 @@ namespace
...
+            if constexpr (error_handling == ErrorHandling::Null)
+                col_null_map = ColumnUInt8::create(input_rows_count, 0);
+            if constexpr (parseDateTime64)
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:652]: No need to use templatization here and blow up the code size needlessly.

We should instead use a pattern like this:

```cpp
if (const auto * result_type = checkAndGetDataType<DataTypeDateTime64>(result_type)
{
    ...
    executeImpl2(...)
}
if (const auto * result_type = checkAndGetDataType<DataTypeDateTime>(result_type)
{
    ...
    executeImpl2(...)
}
```

File: src/Functions/parseDateTime.cpp [hunk 5]
```diff
@@ -598,13 +610,71 @@ namespace
...
+            {
+                const DataTypeDateTime64 * datatime64_type = checkAndGetDataType<DataTypeDateTime64>(removeNullable(result_type).get());
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:654]: `removeNullable` seems arbitrary. Why is it not done for the normal DateTime case? What about `removeSparse()` and `removeLowCardinality`? And why do the SQL tests contain no example with Nullable inputs.

File: src/Functions/parseDateTime.cpp [hunk 6]
```diff
@@ -684,7 +746,16 @@ namespace
                 if (result.has_value())
                 {
                     if (result = datetime.buildDateTime(time_zone); result.has_value())
-                        res_data[i] = static_cast<UInt32>(*result);
+                    {
+                        if constexpr (parseDateTime64)
+                        {
+                            const DataTypeDateTime64 * datatime64_type = checkAndGetDataType<DataTypeDateTime64>(removeNullable(result_type).get());
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:752]: The stuff computed in l. 752-753 doesn't change between rows. We should pull it out of the loop.

File: src/Functions/parseDateTime.cpp [hunk 7]
```diff
@@ -437,6 +439,16 @@ namespace
             return {};
         }
 
+        [[nodiscard]]
+        VoidOrError setMicrosecond(Int32 microsecond_)
+        {
+            if (microsecond_ < 0 || microsecond_ > 999999)
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:445]: 999999 seems arbitrary. Users who specify more than 6 x `S` placeholders will expect nanosecond precision. I think microsecond precision is generally okay but then we should disallow more than 6 x `S` in the format string.

File: src/Functions/parseDateTime.cpp [hunk 8]
```diff
@@ -1625,6 +1693,64 @@ namespace
                 RETURN_ERROR_IF_FAILED(date.setSecond(second))
                 return cur;
             }
+
+            [[nodiscard]]
+            static PosOrError jodaMicroSecondOfSecond(size_t repetitions, Pos cur, Pos end, const String & fragment, DateTime<error_handling> & date)
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:1698]: `MicroSecond` --> `Microsecond`

File: src/Functions/parseDateTime.cpp [hunk 9]
```diff
@@ -598,13 +610,71 @@ namespace
...
+                    for (const auto & ins : instructions)
+                    {
+                        if (scale > 0)
+                            break;
+                        const String fragment = ins.getFragment();
+                        for (char ch : fragment)
+                        {
+                            if (ch != 'S')
+                            {
+                                scale = 0;
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:630]: The code makes the assumption that the format string contains just a single sequence of `S`s. We should at least document this.

File: src/Functions/parseDateTime.cpp [hunk 10]
```diff
@@ -672,7 +734,7 @@ namespace
                 Int64OrError result = 0;
 
                 /// Ensure all input was consumed
-                if (cur < end)
+                if (!parseDateTime64 && cur < end)
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:737]: It is not clear why we no longer check for `parseDateTime64`. Please add a comment.

File: src/Functions/parseDateTime.cpp [hunk 11]
```diff
@@ -1625,6 +1693,64 @@ namespace
...
+                else
+                    RETURN_ERROR(
+                        ErrorCodes::CANNOT_PARSE_DATETIME,
+                        "Unable to parse fragment {} from {} because of unknown sign time zone offset: {}",
+                        fragment,
+                        std::string_view(cur, end - cur),
+                        std::string_view(cur, 1))
+                ++cur;
+
+                Int32 hour;
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:1746]: So is there no sanity check that the read hour and minute are within bounds 0-23 and 0-59?

File: src/Functions/parseDateTime.cpp [hunk 12]
```diff
@@ -2038,6 +2167,9 @@ namespace
             }
             else
             {
+                if (!arguments[1].column || !isColumnConst(*arguments[1].column))
```
Comment (rschu1ze) [inline src/Functions/parseDateTime.cpp:2170]: So does l. 2173 not subsume at least the second check?

**Comment by rschu1ze** [comment]
@KevinyhZou Kindly address the comments in a follow-up PR, thanks.

**Comment by KevinyhZou** [comment]
OK

**Comment by clickgapai** [comment]
Hi @KevinyhZou @hanfei1991 — while reviewing this PR I found the following:

- **Bug (P3):** `setScale` error message uses member variable `scale` instead of parameter `scale_`, showing 0 instead of actual invalid value — https://github.com/ClickHouse/ClickHouse/issues/104651

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**